File: front/src/types/bill.ts
export type BillItemCategory = "starters" | "mains" | "desserts" | "drinks";

export interface BillItem {
  id: string;
  name: string;
  category: BillItemCategory;
  quantity: number;
  unitPrice: number;
  subtotal: number;
}

export interface TaxInfo {
  label: string;
  rate: number;
  amount: number;
}

export interface ServiceCharge {
  label: string;
  rate: number;
  amount: number;
}

export type BillStatus = "open" | "partial" | "paid";

export interface Bill {
  id: string;
  status: BillStatus;
  createdAt: string;
  currency: string;
  items: BillItem[];
  subtotal: number;
  tax: TaxInfo;
  serviceCharge: ServiceCharge;
  total: number;
  amountPaid: number;
  amountRemaining: number;
}

export interface Restaurant {
  id: string;
  name: string;
  logo: string;
  address: string;
}

export interface Table {
  id: string;
  number: number;
  session: string;
}

export interface BillResponse {
  restaurant: Restaurant;
  table: Table;
  bill: Bill;
}

/* ── Checkout flow types ── */

export type TipOption = "none" | "10" | "15" | "20" | "custom";

export interface TipState {
  option: TipOption;
  percentage: number | null;
  amount: number;
}

export type PaymentMethod = "apple_pay" | "google_pay" | "card";

export type PaymentStatus = "idle" | "processing" | "success" | "error";

export interface PaymentState {
  method: PaymentMethod | null;
  status: PaymentStatus;
  errorMessage?: string;
}

export type CheckoutStep =
  | "bill"
  | "split"
  | "tip"
  | "payment"
  | "confirmation";

/* ── Split types ── */

export type SplitMode = "none" | "equal" | "by_item" | "custom";

interface NoSplitState {
  mode: "none";
}

export interface EqualSplitState {
  mode: "equal";
  headCount: number;
  myShare: number;
}

export interface ItemSplitState {
  mode: "by_item";
  claimedItemIds: string[];
  sharedItemIds: string[];
  myShare: number;
}

export interface CustomSplitState {
  mode: "custom";
  myAmount: number;
}

export type SplitState =
  | NoSplitState
  | EqualSplitState
  | ItemSplitState
  | CustomSplitState;

File: backend/src/SplitTheBill.Api/Entities/Bill.cs
namespace SplitTheBill.Api.Entities;

public enum BillStatus
{
    Open,
    Partial,
    Paid,
}

public class Bill
{
    public string Id { get; set; } = string.Empty;
    public string TableSessionId { get; set; } = string.Empty;
    public BillStatus Status { get; set; } = BillStatus.Open;
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public string Currency { get; set; } = "AZN";
    public string TaxLabel { get; set; } = string.Empty;
    public decimal TaxRate { get; set; }
    public string ServiceChargeLabel { get; set; } = string.Empty;
    public decimal ServiceChargeRate { get; set; }

    public TableSession Session { get; set; } = null!;
    public List<BillItem> Items { get; set; } = [];
    public List<Payment> Payments { get; set; } = [];
}
